c7n/tags.py

```diff
@@ -917,6 +920,127 @@ def process_resource_set(self, resource_set, tags):
+        skip_missing={'type': 'boolean'},
+        key={'type': 'string'},
+        tags={'oneOf': [
+            {'enum': ['*']},
+            {'type': 'array'}
+        ]},
+        required=['tags', 'key', 'resource']
+    )
+
+    def get_permissions(self):
+        return self.manager.action_registry.get('tag').permissions
+
+    def validate(self):
+        related_resource = self.data['resource']
+        if related_resource not in all_resources.keys():
+            raise PolicyValidationError(
+                "Error: Invalid resource type selected: %s" % related_resource
+            )
+        # ideally should never raise here since we shouldn't be applying this
+        # action to a resource if it doesn't have a tag action implemented
+        if self.manager.action_registry.get('tag') is None:
+            raise PolicyValidationError(
+                "Error: Tag action missing on resource"
+            )
+        return self
+
+    def process(self, resources):
+        tag_keys = self.data['tags']
+        if isinstance(tag_keys, six.string_types):
+            tag_keys = ['*']
+        related_key = self.data['key']
+        related_type = self.data['resource']
+
+        related_resource_tuple = list(
+            zip(jmespath.search('[].%s' % related_key, resources), resources))
+
+        related_ids = [r[0] for r in related_resource_tuple]
+        related_tag_map = self.get_resource_tag_map(related_type, related_ids)
+        missing_related_tags = list(set(related_ids).difference(related_tag_map.keys()))
+
+        if not self.data.get('skip_missing', True) and missing_related_tags:
+            raise PolicyExecutionError(
+                "Unable to find all %s resources associated with %s" % (related_type, related_ids))
+
+        # rely on resource manager tag action implementation as it can differ between resources
+        tag_action = self.manager.action_registry.get('tag')({}, self.manager)
+        tag_action.id_key = tag_action.manager.get_model().id
+
+        for related, r in related_resource_tuple:
+            if related in missing_related_tags or not related_tag_map[related]:
+                self.log.warning(
+                    'Tags not found for related resource: %s, skipping %s' % (related, r))
+                continue
+            self.process_resource(r, related_tag_map[related], tag_keys, tag_action)
+
+    def process_resource(self, r, related_tags, tag_keys, tag_action):
+        tags = {}
+        resource_tags = {t['Key']: t['Value'] for t in r.get('Tags', [])}
+        if '*' in tag_keys:
+            tags = related_tags
+        else:
+            tags = {k: v for k, v in related_tags.items()
+                    if k in tag_keys and resource_tags.get(k) != v}
+        if not tags:
+            return
+        tags = [{'Key': k, 'Value': v} for k, v in tags.items()]
+        self.log.info('Adding related tags: %s to resource: %s' % (tags, r[tag_action.id_key]))
```

Comment:
debug level please, imagine get log messages for tens of thousands of snapshots.. 

actually I'd consider having this method return a true value for modification, and then doing a log statement for the aggregate in the parent.